rustup-init.ps1

```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/env powershell
+
+# This is just a little script that can be downloaded from the internet to
+# install rustup. It just does platform detection, downloads the installer and
+# runs it.
+
+#XXX: If you change anything here, please make the same changes in setup_mode.rs
+function Write-Usage {
+    Write-Host @'
+rustup-init 1.18.3 (435397f48 2019-05-22)
+The installer for rustup
+
+USAGE:
+    rustup-init.exe [FLAGS] [OPTIONS]
+
+FLAGS:
+    -v, --verbose           Enable verbose output
+    -y                      Disable confirmation prompt.
+        --no-modify-path    Don't configure the PATH environment variable
+    -h, --help              Prints help information
+    -V, --version           Prints version information
+
+OPTIONS:
+        --default-host <default-host>              Choose a default host triple
+        --default-toolchain <default-toolchain>    Choose a default toolchain to install
+'@
+}
+
+function main([AllowEmptyCollection()] [string[]]$Argv = @()) {
+    foreach ($arg in $Argv) {
+        if ("$arg" -eq "-h" -or "$arg" -eq "--help") {
+            Write-Usage
+            return
+        }
+    }
+
+    Install-Rustup $Argv
+}
+
+function Get-Architecture {
+    $platform = [System.Environment]::OSVersion.Platform
+    switch -Wildcard ($platform) {
+        # Linux and Darwin both report as Unix
+        "Unix" {
+            $ostype = uname -s
+            $cputype = uname -m
+            break
+        }
+        "Win*" {
+            $ostype = (Get-WmiObject Win32_OperatingSystem).Caption
+            $cputype = (Get-WmiObject Win32_OperatingSystem).OSArchitecture
+            break
+        }
+        default {
+            Write-Error "could not determine OS platform: $platform"
+        }
+    }
+
+    switch -Wildcard ($ostype) {
+        "Darwin" {
+            $ostype = "apple-darwin"
+            break
+        }
+        "Linux" {
+            $ostype = "unknown-linux-gnu"
+            break
+        }
+        "*Windows*" {
+            $ostype = "pc-windows-msvc"
+            break
+        }
+        default {
+            Write-Error "unrecognized OS type: $ostype"
+        }
+    }
+
+    switch ($cputype) {
+        # Reported by Windows 32-bit operating systems
+        "32-bit" {
+            $cputype = "i686"
+            break
+        }
+        # Reported by Windows 64-bit operating systems
+        "64-bit" {
+            $cputype = "x86_64"
+            break
+        }
+        # Reported by Darin/Linux 64-bit operating systems
+        "x86_64" {
+            break
+        }
+        default {
+            Write-Error "unknown CPU type: $cputype"
+        }
+    }
+
+    $arch = "${cputype}-${ostype}"
+    Write-Debug "Detected arch '$arch'"
+    return "$arch"
+}
+
+# If RUSTUP_UPDATE_ROOT is unset, default it.
+function Get-RustupUpdateRoot {
+    if (Test-Path env:\RUSTUP_UPDATE_ROOT) {
+        return "$env:RUSTUP_UPDATE_ROOT"
```

Comment:
nit: return is un-necessary.  Just output the string.

nit: Wrapping it in quotes is un-necessary.  It's already a string.

Reply:
```
function Get-RustupUpdateRoot {
    if (Test-Path env:\RUSTUP_UPDATE_ROOT) {
        env:RUSTUP_UPDATE_ROOT
    } else {
        "https://static.rust-lang.org/rustup"
    }
}
```

Warning : Watch out for $null values for the ENV as Test-Path will throw

e.g.

```
C:\Source> Test-Path $null
Test-Path : Cannot bind argument to parameter 'Path' because it is null.
At line:1 char:11
+ Test-Path $null
```

Reply:
Actually in retrospect...this whole thing won't work.   `env:RUSTUP_UPDATE_ROOT` returns an object, not a string. To get the value you need to call the `.value` method.  This is better;

```
function Get-RustupUpdateRoot {
    $UpdateRoot = [System.Environment]::RUSTUP_UPDATE_ROOT
    if (($UpdateRoot -ne $null) -and (Test-Path -Path $UpdateRoot)) {
        $UpdateRoot
    } else {
        "https://static.rust-lang.org/rustup"
    }
}
```